### PR TITLE
File search: improve

### PR DESCRIPTION
**Pull request by hius07**
(1) Interrruptable.
(2) Cache results.
(3) "Last file search results" menu item and action.
Closes https://github.com/koreader/koreader/issues/12544.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12549)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]

**Review by NiLuJe — APPROVED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12549)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @hius07)

File: frontend/apps/filemanager/filemanagerfilesearcher.lua [hunk 1]
```diff
@@ -36,11 +36,24 @@ function FileSearcher:registerKeyEvents()
...
 
 function FileSearcher:onShowFileSearch(search_string)
-    local search_dialog
-    local check_button_case, check_button_subfolders, check_button_metadata
+    local search_dialog, check_button_case, check_button_subfolders, check_button_metadata
+    local function _doSearch()
+        local search_str = search_dialog:getInputText()
+        if search_str == "" then return end
+        self.search_string = search_str
+        G_reader_settings:saveSetting("file_search_string", search_str)
```
Comment (poire-z) [inline frontend/apps/filemanager/filemanagerfilesearcher.lua:44]: In ReaderSearch, it's called `fulltext_search_last_search_text`.
Would be nice if these read similar :) ie. `file_search_last_search_text`

File: frontend/apps/filemanager/filemanagerfilesearcher.lua [hunk 2]
```diff
@@ -110,24 +108,49 @@ function FileSearcher:onShowFileSearch(search_string)
...
+    if #self.search_results > 0 then
+        self:onShowSearchResults(not_cached)
     else
         self:showSearchResultsMessage(true)
     end
 end
 
 function FileSearcher:getList()
-    self.no_metadata_count = 0
+    self.no_metadata_count = 0 -- in subprocess
```
Comment (poire-z) [inline frontend/apps/filemanager/filemanagerfilesearcher.lua:147]: `-- in subprocess` reads vague, I don't know what you mean when reading it.

File: frontend/apps/filemanager/filemanagerfilesearcher.lua [hunk 3]
```diff
@@ -227,7 +250,12 @@ function FileSearcher:showSearchResultsMessage(no_results)
     end
 end
 
-function FileSearcher:showSearchResults(results)
+function FileSearcher:onShowSearchResults(not_cached)
+    if (not not_cached and self.search_results == nil) then
```
Comment (poire-z) [inline frontend/apps/filemanager/filemanagerfilesearcher.lua:254]: Parens seems unnecessary (seeing them, it feels you may have made a typo and wanted to have the `not` outside).

**Comment by hius07** [comment]
Can we introduce smth like `G_common = {}` in reader.lua to keep the results cache between FM/Reader?

**Comment by poire-z** [comment]
You can also just store them in the module, ie. as `FileSearcher.last_search_results` (and not `self.last_search_results`). It will be a property of the require'd/cache object, and never dropped.

File: frontend/apps/filemanager/filemanagerfilesearcher.lua (hunk 2 above)
Comment (hius07) [reply frontend/apps/filemanager/filemanagerfilesearcher.lua:146]: The comment explains the following line:
https://github.com/koreader/koreader/blob/b9e6519c97fec4a65b3a19b6a7f179ab5a84a420/frontend/apps/filemanager/filemanagerfilesearcher.lua#L204
It's another `self.no_metadata_count`.

**Review by NiLuJe — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12549)* status: all files reviewed, 3 unresolved discussions (waiting on @hius07 and @poire-z)

File: frontend/apps/filemanager/filemanagerfilesearcher.lua (hunk 2 above)
Comment (poire-z) [reply frontend/apps/filemanager/filemanagerfilesearcher.lua:146]: So, a better comment maybe?:
`-- will be updated in doSearch() with result from subprocess`

File: frontend/apps/filemanager/filemanagerfilesearcher.lua [hunk 4]
```diff
@@ -206,9 +228,16 @@ function FileSearcher:isFileMatch(filename, fullpath, search_string, is_file)
...
 
 function FileSearcher:showSearchResultsMessage(no_results)
-    local text = no_results and T(_("No results for '%1'."), self.search_string)
+    local text = no_results and T(_("No results for '%1'."), FileSearcher.search_string)
     if self.no_metadata_count == 0 then
-        UIManager:show(InfoMessage:new{ text = text })
+        UIManager:show(ConfirmBox:new{
+            text = text,
+            icon = "notice-info",
+            ok_text = _("New search"),
```
Comment (jonnyl2) [inline frontend/apps/filemanager/filemanagerfilesearcher.lua:236]: I would suggest `Edit search` or `Modify search`. "New" implies a blank slate.

Comment (hius07) [reply frontend/apps/filemanager/filemanagerfilesearcher.lua:236]: Any other variants? I don't feel excited with the proposed ones.

Comment (mergen3107) [reply frontend/apps/filemanager/filemanagerfilesearcher.lua:236]: Continue search? Search again? Just a magnifying glass icon?

**Pull request merged**